File: research/experiments/sweep/runs/opt-small-csharp-sonnet45/Optimization/ResultTypesTests.cs
namespace Optimization;

public class ResultTypesTests
{
    [Fact]
    public void DefaultOptions_CreatesDefaults()
    {
        var opts = ResultTypesHelper.DefaultOptions();
        Assert.Equal(1e-8, opts.GradTol);
        Assert.Equal(1e-8, opts.StepTol);
        Assert.Equal(1e-12, opts.FuncTol);
        Assert.Equal(1000, opts.MaxIterations);
    }

    [Fact]
    public void DefaultOptions_AllowsOverrides()
    {
        var opts = ResultTypesHelper.DefaultOptions(o => o.GradTol = 1e-4);
        Assert.Equal(1e-4, opts.GradTol);
        Assert.Equal(1e-8, opts.StepTol);
        Assert.Equal(1e-12, opts.FuncTol);
        Assert.Equal(1000, opts.MaxIterations);
    }

    [Fact]
    public void CheckConvergence_GradientCriterion()
    {
        var opts = ResultTypesHelper.DefaultOptions();
        var reason = ResultTypesHelper.CheckConvergence(1e-9, 0.1, 0.1, 5, opts);
        Assert.IsType<ConvergenceReason.Gradient>(reason);
    }

    [Fact]
    public void CheckConvergence_StepCriterion()
    {
        var opts = ResultTypesHelper.DefaultOptions();
        var reason = ResultTypesHelper.CheckConvergence(0.1, 1e-9, 0.1, 5, opts);
        Assert.IsType<ConvergenceReason.Step>(reason);
    }

    [Fact]
    public void CheckConvergence_FunctionCriterion()
    {
        var opts = ResultTypesHelper.DefaultOptions();
        var reason = ResultTypesHelper.CheckConvergence(0.1, 0.1, 1e-13, 5, opts);
        Assert.IsType<ConvergenceReason.Function>(reason);
    }

    [Fact]
    public void CheckConvergence_MaxIterations()
    {
        var opts = ResultTypesHelper.DefaultOptions();
        var reason = ResultTypesHelper.CheckConvergence(0.1, 0.1, 0.1, 1000, opts);
        Assert.IsType<ConvergenceReason.MaxIterations>(reason);
    }

    [Fact]
    public void CheckConvergence_NoCriterionMet()
    {
        var opts = ResultTypesHelper.DefaultOptions();
        var reason = ResultTypesHelper.CheckConvergence(0.1, 0.1, 0.1, 5, opts);
        Assert.Null(reason);
    }

    [Fact]
    public void IsConverged_SuccessCases()
    {
        Assert.True(ResultTypesHelper.IsConverged(new ConvergenceReason.Gradient()));
        Assert.True(ResultTypesHelper.IsConverged(new ConvergenceReason.Step()));
        Assert.True(ResultTypesHelper.IsConverged(new ConvergenceReason.Function()));
    }

    [Fact]
    public void IsConverged_FailureCases()
    {
        Assert.False(ResultTypesHelper.IsConverged(new ConvergenceReason.MaxIterations()));
        Assert.False(ResultTypesHelper.IsConverged(new ConvergenceReason.LineSearchFailed()));
    }

    [Fact]
    public void ConvergenceMessage_ReturnsReadableMessages()
    {
        Assert.Contains("gradient", ResultTypesHelper.ConvergenceMessage(new ConvergenceReason.Gradient()), StringComparison.OrdinalIgnoreCase);
        Assert.Contains("step", ResultTypesHelper.ConvergenceMessage(new ConvergenceReason.Step()), StringComparison.OrdinalIgnoreCase);
        Assert.Contains("function", ResultTypesHelper.ConvergenceMessage(new ConvergenceReason.Function()), StringComparison.OrdinalIgnoreCase);
        Assert.Contains("maximum iterations", ResultTypesHelper.ConvergenceMessage(new ConvergenceReason.MaxIterations()), StringComparison.OrdinalIgnoreCase);
        Assert.Contains("line search", ResultTypesHelper.ConvergenceMessage(new ConvergenceReason.LineSearchFailed()), StringComparison.OrdinalIgnoreCase);
    }

    [Fact]
    public void CheckConvergence_PriorityOrder()
    {
        // When multiple criteria are met, gradient has priority
        var opts = ResultTypesHelper.DefaultOptions();
        var reason = ResultTypesHelper.CheckConvergence(1e-9, 1e-9, 1e-13, 5, opts);
        Assert.IsType<ConvergenceReason.Gradient>(reason);
    }
}

File: research/experiments/sweep/runs/opt-small-swift-sonnet45/Tests/OptimizationTests/ResultTypesTests.swift
import XCTest
@testable import Optimization

final class ResultTypesTests: XCTestCase {

    // MARK: - defaultOptions

    func testDefaultOptions() {
        let opts = ResultTypes.defaultOptions()
        XCTAssertEqual(opts.gradTol, 1e-8, accuracy: 1e-20)
        XCTAssertEqual(opts.stepTol, 1e-8, accuracy: 1e-20)
        XCTAssertEqual(opts.funcTol, 1e-12, accuracy: 1e-20)
        XCTAssertEqual(opts.maxIterations, 1000)
    }

    func testDefaultOptionsWithOverrides() {
        let opts = ResultTypes.defaultOptions(gradTol: 1e-4)
        XCTAssertEqual(opts.gradTol, 1e-4, accuracy: 1e-20)
        XCTAssertEqual(opts.stepTol, 1e-8, accuracy: 1e-20)
        XCTAssertEqual(opts.funcTol, 1e-12, accuracy: 1e-20)
        XCTAssertEqual(opts.maxIterations, 1000)
    }

    // MARK: - checkConvergence

    func testCheckConvergenceGradient() {
        let opts = ResultTypes.defaultOptions()
        let reason = ResultTypes.checkConvergence(
            gradNorm: 1e-9,
            stepNorm: 0.1,
            funcChange: 0.1,
            iteration: 5,
            options: opts
        )
        XCTAssertEqual(reason, .gradient)
    }

    func testCheckConvergenceStep() {
        let opts = ResultTypes.defaultOptions()
        let reason = ResultTypes.checkConvergence(
            gradNorm: 0.1,
            stepNorm: 1e-9,
            funcChange: 0.1,
            iteration: 5,
            options: opts
        )
        XCTAssertEqual(reason, .step)
    }

    func testCheckConvergenceFunction() {
        let opts = ResultTypes.defaultOptions()
        let reason = ResultTypes.checkConvergence(
            gradNorm: 0.1,
            stepNorm: 0.1,
            funcChange: 1e-13,
            iteration: 5,
            options: opts
        )
        XCTAssertEqual(reason, .function)
    }

    func testCheckConvergenceMaxIterations() {
        let opts = ResultTypes.defaultOptions()
        let reason = ResultTypes.checkConvergence(
            gradNorm: 0.1,
            stepNorm: 0.1,
            funcChange: 0.1,
            iteration: 1000,
            options: opts
        )
        XCTAssertEqual(reason, .maxIterations)
    }

    func testCheckConvergenceNone() {
        let opts = ResultTypes.defaultOptions()
        let reason = ResultTypes.checkConvergence(
            gradNorm: 0.1,
            stepNorm: 0.1,
            funcChange: 0.1,
            iteration: 5,
            options: opts
        )
        XCTAssertNil(reason)
    }

    // MARK: - Priority test

    func testConvergencePriority() {
        // When multiple criteria are met, gradient takes priority
        let opts = ResultTypes.defaultOptions()
        let reason = ResultTypes.checkConvergence(
            gradNorm: 1e-9,
            stepNorm: 1e-9,
            funcChange: 1e-13,
            iteration: 1000,
            options: opts
        )
        XCTAssertEqual(reason, .gradient)
    }

    // MARK: - isConverged

    func testIsConvergedGradient() {
        XCTAssertTrue(ResultTypes.isConverged(.gradient))
    }

    func testIsConvergedStep() {
        XCTAssertTrue(ResultTypes.isConverged(.step))
    }

    func testIsConvergedFunction() {
        XCTAssertTrue(ResultTypes.isConverged(.function))
    }

    func testIsConvergedMaxIterations() {
        XCTAssertFalse(ResultTypes.isConverged(.maxIterations))
    }

    func testIsConvergedLineSearchFailed() {
        XCTAssertFalse(ResultTypes.isConverged(.lineSearchFailed))
    }

    // MARK: - convergenceMessage

    func testConvergenceMessage() {
        XCTAssertEqual(
            ResultTypes.convergenceMessage(.gradient),
            "Converged: gradient norm below tolerance"
        )
        XCTAssertEqual(
            ResultTypes.convergenceMessage(.step),
            "Converged: step size below tolerance"
        )
        XCTAssertEqual(
            ResultTypes.convergenceMessage(.function),
            "Converged: function change below tolerance"
        )
        XCTAssertEqual(
            ResultTypes.convergenceMessage(.maxIterations),
            "Stopped: maximum iterations reached"
        )
        XCTAssertEqual(
            ResultTypes.convergenceMessage(.lineSearchFailed),
            "Stopped: line search failed"
        )
    }
}
